File: src/api/redstone.py
import requests

def get_price_from_redstone(symbol: str) -> float | None:
    symbol = symbol.upper()
    try:
        url = f"https://api.redstone.finance/prices/?symbol={symbol}&provider=redstone&limit=1"
        response = requests.get(url, timeout=5)
        response.raise_for_status()
        data = response.json()

        if isinstance(data, list) and len(data) > 0:
            return data[0]["value"]

        print(f"[RedStone] 沒有取得 {symbol} 的價格資料")
        return None

    except Exception as e:
        print(f"[RedStone] 查詢 {symbol} 價格失敗：{e}")
        return None

if __name__ == "__main__":
    from rich import print
    print(get_price_from_redstone("eth"))

File: src/api/redstone.rs
use reqwest::Client;
use serde::Deserialize;

#[derive(Deserialize, Debug)]
struct RedstonePrice {
    value: f64,
}

pub async fn get_price_from_redstone(symbol: &str) -> Result<f64, String> {
    let symbol = symbol.to_uppercase();
    let url = format!(
        "https://api.redstone.finance/prices/?symbol={}&provider=redstone&limit=1",
        symbol
    );

    let client = Client::builder()
        .timeout(std::time::Duration::from_secs(5))
        .build()
        .map_err(|e| e.to_string())?;

    let response = client.get(&url).send().await.map_err(|e| {
        format!("[RedStone] Failed to query price for {}: {}", symbol, e)
    })?;

    if response.status().is_success() {
        let data: Vec<RedstonePrice> = response.json().await.map_err(|e| {
            format!("[RedStone] Returned JSON format error: {}", e)
        })?;

        if let Some(price_data) = data.get(0) {
            Ok(price_data.value)
        } else {
            Err(format!("[RedStone] No price data found for {}", symbol))
        }
    } else {
        Err(format!("[RedStone] HTTP error code: {}", response.status()))
    }
}
